src/python/py_parsers.cpp

```diff
@@ -0,0 +1,221 @@
+
+namespace {
+
+
+class ConstructableIncrementalParserBase: public IncrementalParserBase {
+public:
+    std::string parse(
+        JsonContainer& msg,
+        const std::string& previous_text, 
+        std::string& delta_text, 
+        const std::optional<std::vector<int64_t>>& previous_tokens = std::nullopt, 
+        const std::optional<std::vector<int64_t>>& delta_tokens = std::nullopt
+    ) override {
+        PYBIND11_OVERRIDE_PURE(
+            std::string,  // Return type
+            IncrementalParserBase,  // Parent class
+            parse,  // Name of function in C++ (must match Python name)
+            msg,
+            previous_text,
+            delta_text,
+            previous_tokens,
+            delta_tokens
+        );
+    }
+};
+
+class ConstructableParserBase: public ParserBase {
+public:
+    void parse(JsonContainer& text) override {
+        PYBIND11_OVERRIDE_PURE(
+            void,  // Return type
+            ParserBase,  // Parent class
+            parse,  // Name of function in C++ (must match Python name)
+            text  // Argument(s)
+        );
+    }
+};
+
+
+// wrapper to enhance calling parser from Python
+void call_parser(py::dict& msg, std::function<void(JsonContainer&)> func) {
+    auto msg_cpp = pyutils::py_object_to_json_container(msg);
+    func(msg_cpp);
+
+    // TODO: msg = pyutils::json_container_to_py_object(msg_cpp) does not work properly here,
+    py::object json_mod = py::module_::import("json");
+    
+    // since it create a new object instead of updating existing dict.
+    auto json_str = msg_cpp.to_json_string();
+    py::dict result = json_mod.attr("loads")(json_str);
+    
+    // update msg with result
+    msg.clear();
+    for (auto item : result) {
+        msg[item.first] = item.second;
+    }
+}
+
+// wrapper to enhance calling incremental parser from Python
+std::string call_incremental_parser(
+    IncrementalParserBase& parser,
+    py::dict& msg,
+    const std::string& previous_text,
+    std::string& delta_text,
+    const std::optional<std::vector<int64_t>>& previous_tokens,
+    const std::optional<std::vector<int64_t>>& delta_tokens,
+    std::function<std::string(JsonContainer&, const std::string&, std::string&, const std::optional<std::vector<int64_t>>&,
+                               const std::optional<std::vector<int64_t>>&)> func) {
+    auto msg_cpp = pyutils::py_object_to_json_container(msg);
+
+    auto res = func(msg_cpp, previous_text, delta_text, previous_tokens, delta_tokens);
+    auto json_str = msg_cpp.to_json_string();
+    
+    // TODO: msg = pyutils::json_container_to_py_object(msg_cpp) does not work properly here,
+    // since it create a new object instead of updating existing dict.
+    py::object json_mod = py::module_::import("json");
+    py::dict result = json_mod.attr("loads")(json_str);
+    // update msg with result
+    msg.clear();
+    for (auto item : result) {
+        msg[item.first] = item.second;
+    }
```

Comment:
Same comment - can we use `pyutils::json_container_to_py_object` instead of explicit json module to create result and then use `update()` method after `msg.clear()`?